src/board.rs

```diff
@@ -248,12 +288,10 @@ impl Board {
     pub fn get_piece_at_position(
         &self,
         square: Square,
-    ) -> Result<Option<(PieceType, Side)>, &'static str> {
-        if !self.is_valid_square(square) {
-            Err("Position out of bounds")
-        } else {
-            Ok(self.squares[square.rank * self.width + square.file])
-        }
+    ) -> Result<Option<(PieceType, Side)>, InvalidSquareError> {
+        self.validate_square(square).map(|square| {
+            self.squares[square.rank * self.width + square.file]
+        }) 
```

Comment:
I like this kind of refactor, it's way more idiomatic.

There's also a question of types. Clearly out of scope for this PR, but it might be a good idea in the future for us to refactor Square into a Square that needs to be checked and one that doesn't (is guaranteed to be correct). Maybe I've been working too much with dependent types but this feels like a validation we shouldn't have to pepper everywhere. Perhaps a lite something like the validation crates that are floating around.

Reply:
That sounds like a good idea. I do really like the idea of "make invalid states unrepresentable"